### PR TITLE
Hawk/dove updating risk attitudes

### PR DESCRIPTION
**Pull request by rlskoeser**
modify variable-risk hawk/dove game to add logic for updating risk levels based on neighbors
- parameter for adjustment strategy: adopt, average, none = no adjustment
- parameter for how often to adjust

adjustment logic is adapted from previous implementation in the risk bet simulation

### CODE REVIEW

File: simulatingrisk/charts/histogram.py [hunk 1]
```diff
@@ -73,4 +73,4 @@ def plot_risk_histogram(model):
     ax.set_title("risk levels")
     # You have to specify the dependencies as follows, so that the figure
     # auto-updates when viz.model or viz.df is changed.
```
Comment (quadrismegistus) [inline simulatingrisk/charts/histogram.py:75]: Is this comment outdated?

File: simulatingrisk/hawkdovevar/app.py [hunk 2]
```diff
@@ -10,8 +10,30 @@
 )
 from simulatingrisk.hawkdove.app import plot_hawks
 
+# adjust single-risk params for variable risk
 jupyterviz_params_var = jupyterviz_params.copy()
+# remove parameter for agent risk level;
```
Comment (quadrismegistus) [inline simulatingrisk/hawkdovevar/app.py:15]: Why is this removed? not clear from immediate context but perhaps clear elsewhere

File: simulatingrisk/hawkdovevar/model.py [hunk 3]
```diff
@@ -13,18 +17,94 @@ def set_risk_level(self):
         # generate a random risk level
         self.risk_level = self.random.randint(0, num_neighbors)
```
Comment (quadrismegistus) [inline simulatingrisk/hawkdovevar/model.py:18]: `self.random` presumably comes from the base mesa class? Otherwise wondering what it refers to

File: simulatingrisk/hawkdovevar/model.py [hunk 4]
```diff
@@ -13,18 +17,94 @@ def set_risk_level(self):
...
+            # adjust risk based on model configuration
+            if self.model.risk_adjustment == "adopt":
+                # adopt neighbor's risk level
+                self.risk_level = best.risk_level
+            elif self.model.risk_adjustment == "average":
+                # average theirs with mine, then round to a whole number
+                # since this model uses discrete risk levels
+                self.risk_level = round(
+                    statistics.mean([self.risk_level, best.risk_level])
+                )
```
Comment (quadrismegistus) [inline simulatingrisk/hawkdovevar/model.py:51]: Very, very cool, and slick. I remember talking about this and/or similar strategies of risk adjustment with David and Lara and you.

File: tests/test_hawkdove.py [hunk 5]
```diff
@@ -67,7 +66,10 @@ def test_agent_repr():
     agent_id = 1
     risk_level = 3
     agent = HawkDoveSingleRiskAgent(agent_id, Mock(agent_risk_level=risk_level))
-    assert repr(agent) == f"<HawkDoveSingleRiskAgent id={agent_id} r={risk_level}>"
+    assert (
+        repr(agent)
+        == f"<HawkDoveSingleRiskAgent id={agent_id} r={risk_level} points=0>"
+    )
```
Comment (quadrismegistus) [inline tests/test_hawkdove.py:72]: nice test

File: tests/test_hawkdovevar.py [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+    assert model.hawk_odds == 0.2
+    assert model.include_diagonals is False
+
+    # handle string none for solara app parameters
+    model = HawkDoveVariableRiskModel(5, risk_adjustment="none")
+    assert model.risk_adjustment is None
+
+    # complain about invalid adjustment type
+    with pytest.raises(ValueError, match="Unsupported risk adjustment 'bogus'"):
+        HawkDoveVariableRiskModel(3, risk_adjustment="bogus")
```
Comment (quadrismegistus) [inline tests/test_hawkdovevar.py:41]: These easy access object property-methods really make using and testing quite easy and legible.

File: tests/test_hawkdovevar.py [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+adjustment_testdata = [
+    # init parameters, expected adjustment round
+    ({"risk_adjustment": None}, None),
+    ({"risk_adjustment": "adopt"}, 10),
+    ({"risk_adjustment": "average"}, 10),
+    ({"risk_adjustment": "average", "adjust_every": 3}, 3),
+]
+
+
+@pytest.mark.parametrize("params,expect_adjust_step", adjustment_testdata)
```
Comment (quadrismegistus) [inline tests/test_hawkdovevar.py:63]: This is nice, so it runs the test 4 times according to the number of tuples in the `adjustment_testdata` list?

File: tests/test_hawkdovevar.py [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+def test_adjust_risk_adopt():
+    # initialize an agent with a mock model
+    agent = HawkDoveVariableRiskAgent(1, Mock(risk_adjustment="adopt"))
+    # set a known risk level
+    agent.risk_level = 2
+    # adjust wealth as if the model had run
+    agent.points = 20
+    # set a mock neighbor with more points than current agent
+    neighbor = Mock(points=1500, risk_level=3)
+    with patch.object(HawkDoveVariableRiskAgent, "most_successful_neighbor", neighbor):
```
Comment (quadrismegistus) [inline tests/test_hawkdovevar.py:121]: very elegant patching

**Review by quadrismegistus — APPROVED** [review]
Fantastic!

File: simulatingrisk/charts/histogram.py (hunk 1 above)
Comment (rlskoeser) [reply simulatingrisk/charts/histogram.py:75]: ooh, yes, thanks for catching

File: simulatingrisk/hawkdovevar/model.py (hunk 3 above)
Comment (rlskoeser) [reply simulatingrisk/hawkdovevar/model.py:18]: yes, mesa provides access to built-in random methods that works just like the main python library random module

It's in mentioned in their documentation best practices, it would allow using a random seed at model instantiation: https://mesa.readthedocs.io/en/stable/best-practices.html#randomization

File: tests/test_hawkdovevar.py (hunk 7 above)
Comment (rlskoeser) [reply tests/test_hawkdovevar.py:63]: yeah, isn't that cool? very useful sometimes
